vllm/multimodal/lmdb_cache.py

```diff
@@ -0,0 +1,657 @@
+
+if TYPE_CHECKING:
+    from multiprocessing.synchronize import Event
+
+    from vllm.config import VllmConfig
+
+    from .cache import MultiModalProcessorCacheInItem, MultiModalProcessorCacheOutItem
+
+logger = init_logger(__name__)
+
+
+class LmdbMultiModalCache:
+    """LMDB-based multi-modal processor cache."""
+
+    CACHES_DIR = os.path.join(envs.VLLM_CACHE_ROOT, "mm_caches")
+    OPEN_ENVS_LOCK = threading.Lock()
+    OPEN_ENVS = dict[str, lmdb.Environment]()
+
+    DB_TIMESTAMPS_AND_HASHES = b"timestamps_and_hashes"
+    DB_HASH_TO_TIMESTAMP = b"hash_to_timestamp"
+    DB_HASH_TO_OBJECT = b"hash_to_object"
+    DB_HASH_TO_PROMPT_UPDATES = b"hash_to_prompt_updates"
+
+    MAP_SIZE_MULTIPLIER = 2
+    MINIMUM_MAP_SIZE = GiB_bytes
+
+    INT_SIZE = 4  # Used for both timestamps (seconds) and chunk indices
+    LMDB_PAGE_HEADER_SIZE = 16
+    LMDB_PAGE_ID_SIZE = 8
+    HASH_ITEM_KEY = "lmdb_mm_hash"
+
+    EVICTOR_READER_CHECK_INTERVAL = 60.0  # seconds
+    EVICTOR_BATCH_SIZE_FRACTION = 0.6  # Relative to the max page ids per page
+    EVICTOR_MIN_UTILIZATION = 0.5  # Start evicting at 50% utilization
+    EVICTOR_MAX_UTILIZATION = 1.0  # Reach maximum duty cycle at 100% utilization
+    EVICTOR_MAX_INTERVAL = 15.0  # seconds
+    EVICTOR_MAX_DUTY_CYCLE = 0.1  # 10%
+
+    def __init__(
+        self,
+        cache_dir: str,
+        cache_size: int,
+        min_eviction_age: int,
+        max_object_size: int,
+    ) -> None:
+        super().__init__()
+        os.makedirs(cache_dir, exist_ok=True)
+
+        self._cache_dir = cache_dir
+        self._cache_size = cache_size
+        self._min_eviction_age = min_eviction_age
+        self._max_object_size = max_object_size
+
+        with self.OPEN_ENVS_LOCK:
+            # A given LMDB environment can only be opened once per process,
+            # and not across forks.
+            if existing_env := self.OPEN_ENVS.get(cache_dir):
+                logger.debug("Reusing existing LMDB environment at %s", cache_dir)
+                self.lmdb_env = existing_env
+            else:
+                # LMDB can require additional space beyond the maximum amoun of data
+                # we're storing, so allocate extra space. (Once the map size is
+                # exhausted, even eviction is liable to fail.)
+                map_size = max(
+                    self._cache_size * self.MAP_SIZE_MULTIPLIER, self.MINIMUM_MAP_SIZE
+                )
+                self.lmdb_env = lmdb.Environment(
+                    path=cache_dir,
+                    map_size=map_size,
+                    max_dbs=4,
+                    writemap=True,
+                    map_async=True,
+                )
+
+                self.OPEN_ENVS[cache_dir] = self.lmdb_env
+
+                # Ensure that the LMDB environment is closed/removed on fork.
+                os.register_at_fork(
+                    after_in_child=lambda: self.OPEN_ENVS.pop(cache_dir).close()
```

Comment:
Replace this with a single registration that clears the dictionary (otherwise the same cache_dir can be registered multiple times in successive forks).